chain/network/src/network_protocol/proto_conv.rs

```diff
@@ -0,0 +1,528 @@
+    type Error = anyhow::Error;
+    fn try_from(x: &GenesisId) -> anyhow::Result<Self> {
+        Ok(Self { chain_id: x.chain_id.clone(), hash: Some((&x.hash).try_into().context("hash")?) })
+    }
+}
+
+//////////////////////////////////////////
+
+impl TryFrom<&proto::PeerChainInfo> for PeerChainInfoV2 {
+    type Error = anyhow::Error;
+    fn try_from(p: &proto::PeerChainInfo) -> anyhow::Result<Self> {
+        Ok(Self {
+            genesis_id: (|| GenesisId::try_from(p.genesis_id.as_ref().context("missing")?))()
+                .context("genesis_id")?,
+            height: p.height,
+            tracked_shards: p.tracked_shards.clone(),
+            archival: p.archival,
+        })
+    }
+}
+
+impl TryFrom<&PeerChainInfoV2> for proto::PeerChainInfo {
+    type Error = anyhow::Error;
+    fn try_from(x: &PeerChainInfoV2) -> anyhow::Result<Self> {
+        Ok(Self {
+            genesis_id: Some((&x.genesis_id).try_into().context("genesis_id")?),
+            height: x.height,
+            tracked_shards: x.tracked_shards.clone(),
+            archival: x.archival,
+        })
+    }
+}
+
+//////////////////////////////////////////
+
+impl TryFrom<&proto::PublicKey> for PeerId {
+    type Error = anyhow::Error;
+    fn try_from(p: &proto::PublicKey) -> anyhow::Result<Self> {
+        Ok(Self::try_from_slice(&p.borsh)?)
+    }
+}
+
+impl TryFrom<&PeerId> for proto::PublicKey {
+    type Error = anyhow::Error;
+    fn try_from(x: &PeerId) -> anyhow::Result<Self> {
+        Ok(Self { borsh: x.try_to_vec()? })
+    }
+}
+
+//////////////////////////////////////////
+
+impl TryFrom<&proto::PartialEdgeInfo> for PartialEdgeInfo {
+    type Error = anyhow::Error;
+    fn try_from(p: &proto::PartialEdgeInfo) -> anyhow::Result<Self> {
+        Ok(Self::try_from_slice(&p.borsh)?)
+    }
+}
+
+impl TryFrom<&PartialEdgeInfo> for proto::PartialEdgeInfo {
+    type Error = anyhow::Error;
+    fn try_from(x: &PartialEdgeInfo) -> anyhow::Result<Self> {
+        Ok(Self { borsh: x.try_to_vec()? })
+    }
+}
+
+//////////////////////////////////////////
+
+impl TryFrom<&PeerInfo> for proto::PeerInfo {
+    type Error = anyhow::Error;
+    fn try_from(x: &PeerInfo) -> anyhow::Result<Self> {
+        Ok(Self { borsh: x.try_to_vec()? })
+    }
+}
+
+impl TryFrom<&proto::PeerInfo> for PeerInfo {
+    type Error = anyhow::Error;
+    fn try_from(x: &proto::PeerInfo) -> anyhow::Result<Self> {
+        Ok(Self::try_from_slice(&x.borsh)?)
+    }
+}
+
+//////////////////////////////////////////
+
+impl TryFrom<&proto::Handshake> for Handshake {
+    type Error = anyhow::Error;
+    fn try_from(p: &proto::Handshake) -> anyhow::Result<Self> {
+        Ok(Self {
+            protocol_version: p.protocol_version,
+            oldest_supported_version: p.oldest_supported_version,
+            sender_peer_id: (|| PeerId::try_from(p.sender_peer_id.as_ref().context("missing")?))()
+                .context("sender_peer_id")?,
```

Comment:
I wonder if this pattern worth making a macro of:

```
required!(p.sender_peer_id)
```

where the macro matches `$id.$id`?

Reply:
Personally I hate macros as their main feature is obscuring the code they evaluate to. However in this particular case I can just define a function like:

fn try_from_required<'a,X,Y:TryFrom<&'a X,Error=anyhow::Error>>(x:&'a Option<X>) -> anyhow::Result<Y> {
    x.as_ref().context("missing")?.try_into()
}